### ISSUE TITLE
Referrential Transparency: Snapkit <> NSLayoutAnchors?

### ISSUE THREAD

**Issue by AndrewSB** [SnapKit/SnapKit#232]
### New Issue Checklist

* [x] I have looked at the [Documentation](http://snapkit.io/docs)
* [x] I have read the [F.A.Q.](http://snapkit.io/faq)

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform                | ios
 Platform Version        | 8.0
 SnapKit Version         | 0.21.0
 Integration Method      | carthage
 

### Issue Description

I was taking a look at NSLayoutAnchors, and trying to understand why SnapKit didn't work the same way. These two snippets of code:

```
stackView.topAnchor.constraintEqualToAnchor(self.scrollView.topAnchor)
stackView.widthAnchor.constraintEqualToConstant(self.view.frame.width)
stackView.centerXAnchor.constraintEqualToAnchor(self.scrollView.centerXAnchor)
```

```
stackView.snp_makeConstraints { make in
    make.top.equalTo(scrollView.topAnchor)
    make.width.equalTo(self.view.frame.width)
    make.centerX.equalTo(scrollView.centerXAnchor)
}
```
don't produce the same result. Though they read almost exactly the same.

I'm planning to debug this further over the weekend, but is there any direction/ guidance you can provide which would explain why NSLayoutAnchor doesn't behave as SnapKit does?

**Comment by robertjpayne** [comment]
@AndrewSB what are the differences you're seeing? My guess here is that `make.centerX` is not necessarily identical to a `centerXAnchor` and especially when dealing with `UIScrollView` I'm not 100% aware of what NSLayoutAnchor does.

If you use a plain view instead of the scroll view do you get the same results?

**Comment by robertjpayne** [comment]
@AndrewSB any chance you can provide more info on this?

**Comment by AndrewSB** [comment]
I've tried enabling and disabling each of the constraints, switching b/w SnapKit and NSLayoutAnchors. 

I've found that the `topAnchor` and `centerXAnchor` don't behave the same way in SnapKit that they do with NSLayoutAnchors. 
I'm pretty sure the `topAnchor` not functioning as expected is due to it being embedded in a `scrollView`. Can you help me understand the difference between `make.centerX` and the `centerXAnchor`?

**Comment by robertjpayne** [comment]
@AndrewSB NSLayoutAnchor is a bit different, I'm not entirely sure how it generates the underlying constraints but we use the NSLayoutConstraint API's directly as anchors are only available on iOS 9.0+.

My guess is you're just seeing a difference between them due to some specifics of auto layout. Top anchors for instance I believe can include space for a navigation bar as can a bottom anchor. Where as NSLayoutConstraint will not.

With scroll views it sounds like it's a difference between the content view and the scroll view. In general you will see some discrepancies and we can't really fix that without breaking tons of apps.
